[NETBEANS-1074] Module Review servletjspapi + web.jspparser

sarveshkesharwani opened the pull request:
- Adding license to binaries-list
- Mavenization of servlet3.1-jsp2.3-api.jar

matthiasblaesing commented on enterprise/servletjspapi/nbproject/project.properties:
```diff
@@ -16,7 +16,7 @@
 
 is.autoload=true
-release.external/servlet3.1-jsp2.3-api.jar=modules/ext/servlet3.1-jsp2.3-api.jar
+release.external/javax.servlet.jsp-api-2.3.3.jar=modules/ext/javax.servlet.jsp-api-2.3.3.jar
```
Please consider if a mapping:
```
release.external/javax.servlet.jsp-api-2.3.3.jar=modules/ext/servlet3.1-jsp2.3-api.jar
```
would be better. That way there would be no external change necessary. You could consider the file name a kind of API and by retaining the original name, you keep the API stable.

That way several of the other changes here are not necessary.

sarveshkesharwani replied:
Matthias, Thanks for the review. This is indeed a better approach to handle this. I have made the necessary changes in the latest commit.

matthiasblaesing commented:
I had another look at this. The current file is created here (seen in your commit):

https://github.com/apache/incubator-netbeans/pull/877/files#diff-b776116ec4c79d0087dad2f3f1fb666a

The file you replace not only holds the JSP api, but also EL and servlet api and schema files. In the referenced build.xml the "instructions" for build the two files (the jsp parser + api) are "documented". You don't need the full glassfish distribution. The EL, Servlet API and JSP API files exist separatedly and the schema files can be extracted from gf-client-module. The latter has a huge number of schemas.

The interesting part will then be determining the final license file. The files look like a mix of CDDL-1.0, CDDL-1.1 and ALv2.

junichi11 commented on enterprise/servletjspapi/external/javax.servlet.jsp-api-2.3.3-license.txt:
```diff
@@ -0,0 +1,833 @@
+License: CDDL-GPL-2-CP
+Description: Java API for Servlets, Java API for JavaServer Pages, Java API for Expression Language
+Origin: Oracle (https://glassfish.dev.java.net/)
+OSR: 5563
```
We can remove this line.

sarveshkesharwani replied:
Thanks Junichi. I have removed this line in the latest commit.

matthiasblaesing commented:
@sarveshkesharwani I think you misunderstood me. Your replacement from maven central is not functional enough. Please have a look at the contents of the library jar before and after your change. See here for the construction of the file:

https://github.com/apache/incubator-netbeans/blob/0190b0ca42f12b358e5f791d6dc2415bfc17bae4/enterprise/web.jspparser/build.xml#L95

That could be integrated into the build (if you don't feel comfortable with it, please indicate so, then I'll have a look at the two modules).

For the license: Oracle/Sun released many files/packages dual-licensed under CDDL-1.X _and_ GPLv2-CPE. At this time GPLv2-CPE is not accepted by the ASF, so when we include such an artifact, we elect to include it under the CDDL license. That is also reflected in the license file, which then only refers to the choosen license.

sarveshkesharwani commented:
@matthiasblaesing , I think I did misunderstand you. There was a lot more going on here. I see that the old jar contained Jsp Api, Servlet Api, El Api, and schema files. I have added the EL, JSP, Servlet and JSP Api jars separately and added licenses for them. Does this look like the right way to approach this?

matthiasblaesing commented:
The idea is sound, but it needs integration into the build system. The "old way" was to build the redistribution jars once and cache them on the binary server. I would not do that anymore. In the last days we learnt the hard way, that we should not rely on external systems (yes I know maven central belongs to that group).

Here is what I would do: Cover `servletjspapi` and `web.jspparser` in one PR (#906 is not enough) as they are both build similar and belong tightly together.

You already identified the API modules, but they need integration into the build process. Please see `enterprise/web.core.syntax` for a similar case. In that module the taglibs are downloaded from maven central and the documentation, that is then redistributed with netbeans, is build as `external/generated-*` (see `prepare-doc` target).

For `servletjspapi` you already identified the source file from maven central. These need to be filtered/unified as done in the `web.jspparser/build.xml` and a grouped license file needs to be created as in the reference module.

The sources for `web.jspparser` module are also available on maven central:
Jasper/JSP API: https://search.maven.org/artifact/org.glassfish.web/javax.servlet.jsp/2.3.4/jar
EL: https://search.maven.org/artifact/org.glassfish.web/javax.el/2.2.6/jar
Preprocessor-Util: https://search.maven.org/artifact/org.glassfish.main.common/common-util/5.0/glassfish-jar
ClassLoaderUtil: https://search.maven.org/artifact/org.glassfish.main.common/glassfish-ee-api/5.0/glassfish-jar

These jars need to be recombined and produces also a `generated*` file. 

The parser has unittests and I ask you to run them before and after your changes, to check that the parser still works.

matthiasblaesing commented:
@sarveshkesharwani we are heading into the right direction. For your questions:

1. The license for the generated files follows the licenses used for the input sources. As a basic rule: If you don't have a license, you can't distribute it. In this case to be faithful to the input, I would declare that the combined files contain CDDL-1.0 and CDDL-1.1 licenses files -- this is true for the API, for the Parser I think I saw only CDDL-1.1.
2. I think your observation is correct the target `create-jars-from-glassfish` is not executed automaticly. I would wager that this is an artifact of the times when maven central was not that popular/when the artifacts were not distributed on maven central. For the approach to regenerate the file every time the build is run: please don't. In `enterprise/web.core.syntax` you can see two thinks: the file is only generated when necessary (`-check-prepared-doc`) and only build when the release target is invoked (`<target name="-process.release.files" depends="prepare-doc"/>`). This can be adapted for these two modules. For `web.jspparser`, `ext-compile` looks like a good place to hook into and for the api the approach from `web.core.syntax` should be usable. I would remove the whole `create-jars-from-glassfish`  when you are done as I gets replaced by your work.
3. Yes tests are pretty much broken. I assume when the donation was prepared `project3` and other files were not deemed donateable and thus removed from the codebase. You can recover at least parts of the unittest by commenting out line 108 in `org.netbeans.modules.web.jspparser.TestUtil#setup`. Then you would at least have a smoke test if something broke.

Further comments:
1. You'll need to revisit the construction of the `servlet3.1-jsp2.3-api.jar` file. You are packaging all schema files, while the original build only choose a subset (see `web.jspparser/build.xml` from line 138 onwards.
2. Please follow the naming established in `web.core.syntax` for the generated artifacts: 1. prefix the filename with `generated-` and add a header to the license file `Type: generated`. Background: the `verify-all-libs-and-licenses` target of the netbeans build will flag the missing files for the licenses - but with the changes suggested in the answer 2. above the file will only be created when needed, so it would be a false positive. See for example `generated-jsf12-tlddoc-1.2-20-license.txt`

sarveshkesharwani commented:
@matthiasblaesing , I have made the necessary changes in the latest commits. I have used an approach similar to `prepare-doc` in `web.core.syntax`  for both servletjspapi and web.jspparser, and I have fixed the licenses for the generated jars as well.
I went with CDDL-1.1 for web.jspparser and a combination of CDDL-1.1 + 1.0 for servletjspapi. I created a new file CDDL-1.0+1.1 for this purpose (because `verify-all-libs-and-licenses` matches the license name). Does this look like the right approach to handle that?

matthiasblaesing commented:
Looks mostly sane. The build process looks ok - the license verification is not yet happy:

- generated-servlet3.1-jsp2.3-api-license.txt is missing "Version" and "Type: generated" headers
- There are license files with trailing spaces (behind the "Oracle"):
enterprise/web.jspparser/external/common-util-5.0-license.txt has a trailing space on line #5                                                                                        
enterprise/web.jspparser/external/glassfish-ee-api-4.1-license.txt has a trailing space on line #5
enterprise/web.jspparser/external/javax.el-2.2.6-license.txt has a trailing space on line #5
enterprise/web.jspparser/external/javax.servlet.jsp-2.3.2-license.txt has a trailing space on line #5
enterprise/servletjspapi/external/el-api-2.2-license.txt has a trailing space on line #5
enterprise/servletjspapi/external/javax.servlet.jsp-api-2.3.3-license.txt has a trailing space on line #5

sarveshkesharwani commented:
@matthiasblaesing , I have 
- added 'Version' and 'Type-generated' headers in generated-servlet3.1-jsp2.3-api-license.txt
- removed the trailing spaces from the said license files.
But the license verification has this to say about almost all the license files : 
`license.txt contains a license body which does not match that in nbbuild/licenses/CDDL-1.1: mismatch around: ' 1. Definitions. 1.1' vs. '&#13; &#13; 1. Definitions.&#13;`
I noticed that this error is displayed for many of the license files across many modules. It's related to the use of space instead of `/r` in the license file. Is there something that should be done about this?

rtaneja1 commented:
>>license.txt contains a license body which does not match that in nbbuild/licenses/CDDL-1.1: mismatch around: ' 1. Definitions. 1.1' vs. '&#13; &#13; 1. Definitions.&#13;
Copy license text from nbbuild/licenses/CDDL-1.1 :)

sarveshkesharwani commented:
> > > Copy license text from nbbuild/licenses/CDDL-1.1 :)

@rtaneja1 , I did that, but the error still won't go away.

matthiasblaesing commented:
Minor nitpick: `Version 2.3` for `generated-servlet3.1-jsp2.3-api-license.txt` looks strange. You might introduce an "artificial" version - for example `generated-servlet-jsp-api-3.1_2.3-license.txt`.

I'm willing to merge as-is please indicate either way. Thank you!

sarveshkesharwani commented:
> I suspect, that you are suffering from autoclrf conversion. The carriage-return (`\r`) is only present on windows and os x. See here:

@matthiasblaesing , You're right. I tried running verification in linux, and the error was no longer present. 

> Minor nitpick: `Version 2.3` for `generated-servlet3.1-jsp2.3-api-license.txt` looks strange. You might introduce an "artificial" version - for example `generated-servlet-jsp-api-3.1_2.3-license.txt`.

I have changed the generated jar's name to `generated-servlet-jsp-api-3.1_2.3.jar` and changed the license name and version in it accordingly.

Thank you so much for all your help. 😃

matthiasblaesing commented:
Thank you for your work. I merged your work with a slight modification in 1f406aead87fd3d679645572db19da5e8a51acbf (yes I screwed up the commit message, missing the web.jspparser). The slight modification was a rename of `generated-servlet-jsp-api-3.1_2.3.jar-license.txt` to `generated-servlet-jsp-api-3.1_2.3-license.txt`.
